### ISSUE TITLE
Quick question about learning rate, batch size and epoch with Horovod

### ISSUE THREAD

**Issue by hoangcuong2011** [horovod/horovod#1617]
**Environment**:

    Framework: (TensorFlow, Keras, PyTorch, MXNet) TensorFlow Keras
    Framework version: 1.15.0
    Horovod version: 0.18.2
    MPI version: openmpi-4.0.2
    CUDA version: 10.0
    NCCL version: 2.5.6
    Python version: 3.6
    OS and version: Ubuntu
    GCC version: 5.4.0

First of all - I would like to give my thanks for providing this amazing library. I just have a quick question regarding to learning rate, batch size and number of epochs with Horovd.

Let us use the [tensorflow keras minist code](https://github.com/horovod/horovod/blob/master/examples/tensorflow_keras_mnist.py) as an example. With this I noticed we have three different lines as follows:

```
model.fit(x_train, y_train,
          batch_size=batch_size,
          callbacks=callbacks,
          epochs=epochs,
          verbose=1 if hvd.rank() == 0 else 0,
          validation_data=(x_test, y_test))
```
`epochs = int(math.ceil(12.0 / hvd.size()))`

`opt = keras.optimizers.Adadelta(1.0 * hvd.size())`

Also let us assume we have 4 gpus to run.
My question: 

1. Given batch_size=batch_size, which is 128 in this example - how the batch_size each instance from these four 4 gpus will process? I suppose it is batch size / num of gpus = 128/4 = 32, but I am not sure it is correct. Could you confirm this (plus any code to confirm this if possible? I could not find this).

2. If we split the batch size to 4, I think the total number of epochs should be still the same - why do we need to divide the number of epochs to 4 as well? 

3. Why do we have to multiply the learning rate with 4? I know it is about adjusting the LR, but I could not find a reason for this. Any detail about this is more than appreciated.

Thank you very much for this!

**Comment by hoangcuong2011** [comment]
After some more investigation (with [this](https://github.com/horovod/horovod/issues/241) and [this](https://github.com/horovod/horovod/issues/699)) - I think I can answer 1 - if we set the batch size of 128, then the full batch size of training is 128*4, not 128/4.

But it is not clear to me with 2 and 3, though.

**Comment by tgaddair** [comment]
Hey @hoangcuong2011, thanks for trying out Horovod.  Let me take a shot at answering your questions.

1. Yes, the batch size will be 128*4.  In Horovod we typically think in terms of "local batch size" and "effective batch size".  Local batch size is whatever you set it to in your training script, effective batch size is the "true" batch size in terms of its effect on model convergence.  If you were to run 4 processes with local batch size 128, it should produce the same results as running a single process with batch size 512.
2. In the MNIST example, we don't split the data between the different workers, we randomly sample from it each step.  Because every batch is a totally random sample, there isn't (mathematically) any difference between each epoch.  So reducing the number of epochs is equivalent to reducing the number of steps per epoch.  When you're partitioning your data instead of randomly sampling it, it would make more sense to reduce the number of `steps_per_epoch` as shown in [keras_imagenet_resnet50.py](https://github.com/horovod/horovod/blob/master/examples/keras_imagenet_resnet50.py#L166).
3. When you increase the batch size, you need to scale the learning rate proportionately to get similar convergence.  If I were running locally with batch size 32, then wanted to double it to 64, I would need to double my learning rate.  Because increasing the number of workers is equivalent to increasing the batch size (see (1)), we need to scale the learning rate by the number of workers.

**Comment by hoangcuong2011** [comment]
Many thanks for this very detailed answer @tgaddair !
